libraries/AP_Relay/AP_Relay.h

```diff
@@ -55,16 +57,25 @@ class AP_Relay {
 
     bool send_relay_status(const class GCS_MAVLINK &link) const;
 
+    void set(AP_Relay_Params::Function function, bool value);
+
+    void update();
+
 private:
     static AP_Relay *singleton;
 
-    AP_Int8 _pin[AP_RELAY_NUM_RELAYS];
-    AP_Int8 _default;
+    AP_Relay_Params _params[AP_RELAY_NUM_RELAYS];
     uint8_t _pin_states;
```

Comment:
i think we should consider removing _pin_states and use the hardware state form hal.gpio, to avoid race conditions between threads

Reply:
There is no longer any internal state retained, all writes are immediately effective, and logging is immediately done as well.